### PR TITLE
Add distributed local cache for StackExchange.Redis provider

### LINKED ISSUE #69: Third Level Cache is it possible
Hi @maca88, 
I have been doing lots of performance test I find out some issues and report with second level cache usage, we are good point now against starting point but i have an idea thinking about third level cache?

Let me tell you i wan't to put in app memory cache on front of redis cache think about that.
Pre request;
1. StandartQueryCache always sync with in memory and redis 
2. UpdateTimestampsCache sync with just redis
3. Object cache always sync with in memory and redis
4. UpdateTimestampsCache never expires

Query<T>().WithOptions(x=>x.SetCacheable(true)).ToList()
Get StandartQueryCache from local cache if exists if not get from redis
Get UpdateTimestamsCache from redis
Compare objects is outdated? 
 a.if outdated it will get result from database and put updated cache to inmemory and redis
 b.if its not outdated it will gets the cache item from in memory cache

So we can **minimize** redis **network cost serialize/deserialize** costs 

image

```c#

public class RedisCacheImpl : RedisCache
{
    public override bool PreferMultipleGet => true;

    public RedisCacheImpl(string regionName, AbstractRegionStrategy regionStrategy) : base(regionName, regionStrategy)
    {

    }
}

public partial class RedisInMemoryCache : RedisCacheImpl
{
    protected CacheBase InMemoryCache { get; private set; }

    public RedisInMemoryCache(string regionName, AbstractRegionStrategy regionStrategy, CacheBase inMemoryCache) 
        : base(regionName, regionStrategy)
    {
        InMemoryCache = inMemoryCache;
    }

    public override object Get(object key)
    {
        var result = InMemoryCache.Get(key);
        if(result == null)
        {
            result = base.Get(key);
            if (result != null)
            {
                InMemoryCache.Put(key, result);
            }
        }
        return result;
    }

    public override object[] GetMany(object[] keys)
    {
        var objects = new object[keys.Length];
        var nullValues = new List<(int index, object key)>();
        for (var i = 0; i < keys.Length; i++)
        {
            var result = InMemoryCache.Get(keys[i]);
            objects[i] = result;
            if(result == null)
            {
                nullValues.Add((i, keys[i]));
            }
        }

        if (nullValues.Count > 0)
        {
            var remoteObjects = base.GetMany(nullValues.Select(x=>x.key).ToArray());
            if(remoteObjects.Length > 0)
            {
                for (int i = 0; i < remoteObjects.Length; i++)
                {
                    var nullValue = nullValues[i];
                    var remoteObject = remoteObjects[i];
                    if (remoteObject != null)
                    {
                        objects[nullValue.index] = remoteObject;
                        InMemoryCache.Put(nullValue.key, remoteObject);
                    }
                }
            }
        }

        return objects;
    }

    public override void Put(object key, object value)
    {
        InMemoryCache.Put(key, value);
        base.Put(key, value);
    }

    public override void PutMany(object[] keys, object[] values)
    {
        for (int i = 0; i < keys.Length; i++)
        {
            InMemoryCache.Put(keys[i], values[i]);
        }
        base.PutMany(keys, values);
    }

    public override void Remove(object key)
    {
        InMemoryCache.Remove(key);
        base.Remove(key);
    }

    public override void Clear()
    {
        InMemoryCache.Clear();
        base.Clear();
    }

    public override void Destroy()
    {
        InMemoryCache.Destroy();
        // No resources to clean-up for redis
    }
}

public partial class RedisInMemoryCache : RedisCacheImpl
{
    public override Task<object> GetAsync(object key, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled<object>(cancellationToken);
        }
        return InternalGetAsync();

        async Task<object> InternalGetAsync()
        {
            var result = await InMemoryCache.GetAsync(key,cancellationToken).ConfigureAwait(false);
            if (result == null)
            {
                result = await base.GetAsync(key,cancellationToken).ConfigureAwait(false);
                if (result != null)
                {
                    await InMemoryCache.PutAsync(key, result,cancellationToken).ConfigureAwait(false);
                }
            }
            return result;
        }
    }

    public override Task<object[]> GetManyAsync(object[] keys, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled<object[]>(cancellationToken);
        }

        return InternalGetManyAsync();

        async Task<object[]> InternalGetManyAsync()
        {
            var objects = new object[keys.Length];

            var nullValues = new List<(int index, object key)>();
            for (var i = 0; i < keys.Length; i++)
            {
                var result = await InMemoryCache.GetAsync(keys[i], cancellationToken).ConfigureAwait(false);
                objects[i] = result;
                if (result == null)
                {
                    nullValues.Add((i, keys[i]));
                }
            }

            if (nullValues.Count > 0)
            {
                var remoteObjects = await base.GetManyAsync(nullValues.Select(x => x.key).ToArray(), cancellationToken).ConfigureAwait(false);
                if (remoteObjects.Length > 0)
                {
                    for (int i = 0; i < remoteObjects.Length; i++)
                    {
                        var nullValue = nullValues[i];
                        var remoteObject = remoteObjects[i];
                        if (remoteObject != null)
                        {
                            objects[nullValue.index] = remoteObject;
                            await InMemoryCache.PutAsync(nullValue.key, remoteObject, cancellationToken).ConfigureAwait(false);
                        }
                    }
                }
            }

            return objects;
        }
    }

    public override Task PutAsync(object key, object value, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }
        return InternalPutAsync();

        async Task InternalPutAsync()
        {
            await InMemoryCache.PutAsync(key, value,cancellationToken).ConfigureAwait(false);
            await base.PutAsync(key, value, cancellationToken).ConfigureAwait(false);
        }
    }

    public override Task PutManyAsync(object[] keys, object[] values, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }
        return InternalPutManyAsync();

        async Task InternalPutManyAsync()
        {
            for (int i = 0; i < keys.Length; i++)
            {
                await InMemoryCache.PutAsync(keys[i], values[i], cancellationToken).ConfigureAwait(false);
            }
            await base.PutManyAsync(keys, values, cancellationToken).ConfigureAwait(false);
        }
    }

    public override Task RemoveAsync(object key, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }
        return InternalRemoveAsync();

        async Task InternalRemoveAsync()
        {
            await InMemoryCache.RemoveAsync(key, cancellationToken).ConfigureAwait(false);
            await base.RemoveAsync(key, cancellationToken).ConfigureAwait(false);
        }
    }

    public override Task ClearAsync(CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }
        return InternalClearAsync();

        async Task InternalClearAsync()
        {
            await InMemoryCache.ClearAsync(cancellationToken).ConfigureAwait(false);
            await base.ClearAsync(cancellationToken).ConfigureAwait(false);
        }
    }
}

public class RedisInMemoryCacheProvider : RedisCacheProvider
{
    private static readonly CoreMemoryCacheProvider memoryCacheProvider = new CoreMemoryCacheProvider();

    private static readonly FieldInfo connectionMultiplexer = typeof(RedisCacheProvider)
        .GetField("_connectionMultiplexer", BindingFlags.NonPublic | BindingFlags.Instance);

    private const string UpdateTimestampsCache = "UpdateTimestampsCache";
    private const string InMemoryUsage = "cache.inmemory_usage";
    private const string InMemoryExpiration = "cache.inmemory_expiration";
    private const string InMemoryExpirationScanFrequency = "cache.inmemory_expiration_scan_frequency";

    protected override CacheBase BuildCache(RedisCacheRegionConfiguration regionConfiguration, IDictionary<string, string> properties)
    {
        var regionStrategy =
            CacheConfiguration.RegionStrategyFactory.Create(connectionMultiplexer.GetValue(this) as IConnectionMultiplexer,
            regionConfiguration, properties);

        var inMemoryUsage = PropertiesHelper.GetBoolean(InMemoryUsage, properties, false);

        if (!inMemoryUsage || 
            regionConfiguration.RegionName.EndsWith(UpdateTimestampsCache))
        {
            return new RedisCacheImpl(regionConfiguration.RegionName, regionStrategy);
        }
        else
        {
            var clonedProperties = properties.ToDictionary(x => x.Key, x => x.Value);
            SetConfiguration(clonedProperties);
            var inMemoryCache = memoryCacheProvider.BuildCache(regionConfiguration.RegionName, clonedProperties) as CacheBase;
            return new RedisInMemoryCache(regionConfiguration.RegionName, regionStrategy, inMemoryCache);
        }
    }

    protected void SetConfiguration(IDictionary<string, string> properties)
    {
        if (CoreMemoryCacheProvider.ExpirationScanFrequency == null)
        {
            var seconds = PropertiesHelper.GetInt32(InMemoryExpirationScanFrequency, properties, 60);
            CoreMemoryCacheProvider.ExpirationScanFrequency = TimeSpan.FromSeconds(seconds);
        }

        const string expiration = "expiration";
        if (!properties.ContainsKey(expiration))
        {
            properties[expiration] = PropertiesHelper.GetInt32(InMemoryExpiration, properties, 60).ToString();
        }
    }
}

```

### PR DESCRIPTION
**Pull request by maca88**
This PR adds three strategies that use a local cache for faster readings, fixes #69.

1. `DistributedLocalCacheRegionStrategy`: Uses only Redis pubsub mechanims to synchronize data between caches.
2. `TwoLayerCacheRegionStrategy`: Uses `DefaultRegionStrategy` logic for updating the Redis cache and Redis pubsub mechanism to invalidate local caches.
3. `FastTwoLayerCacheRegionStrategy`: Uses `FastRegionStrategy` logic for updating the Redis cache and Redis pubsub mechanism to invalidate local caches.

All three strategies share the same drawbacks:
- No batching for `PutMany` operation
- Higher CPU usage due to processing of synchronization/invalidation messages for write operations
- Slow write operations

Additional drawbacks for `DistributedLocalCacheRegionStrategy`:
- Slow Lock/LockMany operation
- After adding/restarting an instance the performance will drop until the cache is populated.

All three strategies supports [pipelining](https://stackexchange.github.io/StackExchange.Redis/PipelinesMultiplexers#pipelining) for write operations, but it is disabled by default as it requires more CPU and may cause a timeout due to the [cpu or network bound](https://stackexchange.github.io/StackExchange.Redis/Timeouts#are-you-getting-network-or-cpu-bound).

The below tests were ran in .NET 4.6.1 on Intel i7-860 and Redis server was located on a different local machine. `StackExchange.Redis` version `2.0.601` was used as it has some performance improvements. In the test environment the bottleneck was the cpu where the test were ran (Intel i7-860) and Redis server cpu when pipelining was used. When enabling pipelining, there were very high synchronization times and also timeouts due to bottlenecking cpu.

| Test method  | &nbsp;TwoLayerCacheRegionStrategy&nbsp; | DistributedLocalCacheRegionStrategy | DefaultRegionStrategy |
| ------------- | -------------- | ------------- | ------------- |
| StressTestGetAsync(1000)  | 520161 op/s  | 2234093 op/s | 7450 op/s |
| StressTestGetManyAsync(1000, 10)  | 72375 op/s  | 1106187 op/s | 7270 op/s |
| StressTestLockUnlockAsync(1000)  | 4265 op/s **\(*)**  | 1171 op/s **\(*)** | 3764 op/s |
| StressTestLockUnlockManyAsync(1000, 10)  | 3929 op/s **\(*)** | 1080 op/s **\(*)** | 4281 op/s |
| StressTestPutAsync(1000)  | 3164 op/s, 11ms sync time **\(*)**<br>With pipelining:<br>33285 op/s, 296ms sync time **\(*)** | 3127 op/s, 2ms sync time **\(*)**<br>With pipelining:<br>31243 op/s, 4143ms sync time **\(*)** | 7211 op/s |
| StressTestPutManyAsync(1000, 10)  | 293 op/s, 13ms sync time **\(*)**<br>With pipelining:<br>1366 op/s, 25ms sync time **\(*)** | 316 op/s, 2ms sync time **\(*)**<br>With pipelining:<br>1170 op/s, 80ms sync time **\(*)**  | 6729 op/s |
| StressTestPutAndGetAsync(1000, 20, 80)  | 6076 op/s, 10ms sync time **\(*)** <br>With pipelining:<br>6804 op/s, 10ms sync time **\(*)** | 15519 op/s, 1ms sync time **\(*)**<br>With pipelining:<br>138007 op/s, 2075ms sync time **\(*)** | 7379 op/s |
| StressTestPutAndClearAsync(1000, 99, 1)  | 3161 op/s, 15ms sync time **\(*)**<br>With pipelining:<br>12648 op/s, 23ms sync time **\(*)** | 3364 op/s, 2ms sync time **\(*)**<br>With pipelining:<br>24894 op/s, 4ms sync time **\(*)** | 6947 op/s |
| StressTestPutAndRemoveAsync(1000, 80, 20)  | 3225 op/s, 11ms sync time **\(*)**<br>With pipelining:<br>33861 op/s, 278ms sync time **\(*)** | 3158 op/s, 2ms sync time **\(*)**<br>With pipelining:<br>32417 op/s, 4126ms sync time **\(*)** | 7346 op/s |
| StressTestPutGetAndClearAsync(1000, 19, 80, 1)  | 5831 op/s, 10ms sync time **\(*)**<br>With pipelining:<br>6309 op/s, 10ms sync time **\(*)** | 18407 op/s, 2ms sync time **\(*)**<br>With pipelining:<br>112632 op/s, 2ms sync time **\(*)** | 6141 op/s |
| StressTestPutGetAndRemoveAsync(1000, 20, 75, 5)  | 5865 op/s, 11ms sync time **\(*)**<br>With pipelining:<br>7009 op/s, 12ms sync time **\(*)** | 16771 op/s, 1ms sync time **\(*)**<br>With pipelining:<br>142914 op/s, 2074ms sync time **\(*)** | 7485 op/s |

**\(*)**  - Only 4 tasks were used in the test, due to the high CPU usage of the strategy.

For tests that do not have **\(*)**, 8 tasks were used.

I didn't realized that the cache providers weren't disposed per test, which heavily impacted the above benchmarks. I've rerun the tests and updated the values.

### CODE REVIEW

**Comment by maca88** [comment]
I didn't see that the cache providers weren't disposed per test, which heavily impacted the above benchmarks. Now the cache providers are disposed per test for Redis provider and I've reran the tests and updated the benchmarks.

**Comment by maca88** [comment]
The last commit contains also a fix for `DefaultRegionStrategy` clear operation as it was possible that the current version was updated to an older one, due to the delay of pubsub messages. Currently, there is an issue with the async generator, which does not generate the cancellation token parameter for `AbstractRegionStrategy.ExecutePutManyAsync`, so I will leave it as WIP until the issue is resolved.

File: StackExchangeRedis/NHibernate.Caches.StackExchangeRedis/RegionMemoryCacheBase.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		/// <summary>
+		/// Puts the value in the cache.
+		/// </summary>
+		/// <param name="key">The key where the value will be store.</param>
+		/// <param name="dependencyKey">The key which <paramref name="key"/> depends on.</param>
+		/// <param name="value">The value to store.</param>
+		public abstract void Put(string key, string dependencyKey, object value);
+
+		/// <summary>
+		/// Puts the value in the cache by using an expicit expiration. When <see cref="TimeSpan.Zero"/>
```
Comment (fredericDelaporte) [inline StackExchangeRedis/NHibernate.Caches.StackExchangeRedis/RegionMemoryCacheBase.cs]: `expicit`

File: StackExchangeRedis/NHibernate.Caches.StackExchangeRedis/ConfigurationHelper.cs [hunk 2]
```diff
@@ -39,6 +39,15 @@ public static int GetInteger(string key, IDictionary<string, string> properties,
 			return properties.TryGetValue(key, out var value) ? Convert.ToInt32(value) : defaultValue;
 		}
 
+		public static long GetLong(string key, IDictionary<string, string> properties, int defaultValue)
```
[the remark this answers is not in the document]
Comment (maca88) [reply StackExchangeRedis/NHibernate.Caches.StackExchangeRedis/ConfigurationHelper.cs]: Yes it should be, thanks for the correction.

File: StackExchangeRedis/NHibernate.Caches.StackExchangeRedis.Tests/CacheRegionStrategyFactory.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Collections.Generic;
+using NHibernate.Caches.StackExchangeRedis.Tests.MemoryCaches;
+using StackExchange.Redis;
+
+namespace NHibernate.Caches.StackExchangeRedis.Tests
+{
+	public class CacheRegionStrategyFactory : DefaultCacheRegionStrategyFactory
```
Comment (fredericDelaporte) [inline StackExchangeRedis/NHibernate.Caches.StackExchangeRedis.Tests/CacheRegionStrategyFactory.cs:8]: Was there a reason not to add support for these new strategies directly in the `DefaultCacheRegionStrategyFactory` @maca88 ?

See #123

Comment (fredericDelaporte) [reply StackExchangeRedis/NHibernate.Caches.StackExchangeRedis.Tests/CacheRegionStrategyFactory.cs:8]: Looking to include it in the default factory, I have now seen the reason, see #123.

**Pull request merged**